pkg/resource/deploy/plan_executor.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploy
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/util/logging"
+)
+
+const (
+	// Utility constant for easy debugging.
+	planExecutorLogLevel = 4
+)
+
+var (
+	// ErrPreviewFailed is returned whenever a preview fails.
+	ErrPreviewFailed = errors.New("preview failed")
+
+	// ErrUpdateFailed is returned whenever an update fails.
+	ErrUpdateFailed = errors.New("update failed")
+
+	// ErrCanceled is returned whenever a plan is canceled.
+	ErrCanceled = errors.New("plan canceled")
+)
+
+// PlanExecutor is responsible for taking a plan and driving it to completion.
+// Its primary responsibility is to own a `stepGenerator` and `stepExecutor`, serving
+// as the glue that links the two subsystems together.
+type PlanExecutor struct {
+	plan    *Plan          // The plan that we are executing
+	opts    Options        // Options for the plan execution
+	src     SourceIterator // The iterator that generates SourceEvents
+	preview bool           // are we running a preview?
+
+	stepGen  *stepGenerator // step generator owned by this plan
+	stepExec *stepExecutor  // step executor owned by this plan
+
+	parentCtx context.Context    // cancellation context for the current CLI session.
+	ctx       context.Context    // cancellation context for the current plan. Child of parentCtx.
+	cancel    context.CancelFunc // CancelFunc that cancels the above context.
+
+	sawError  atomic.Value // have we seen an error?
```

Comment:
Can you file an issue to clean up the error handling here and in the executor?

Reply:
yep, will do.